SQL: Allow sorting of groups by aggregates

costin opened the pull request:
Introduce client-side, custom sorting of groups based on aggregate
functions. To allow this, the Analyzer has been extended to push down
to underlying Aggregate, aggregate function and the Querier has been
extended to identify the case and consume the results in order and sort
them based on the given columns.
The underlying QueryContainer has been slightly modified to allow a view
of the underlying values being extracted as the columns used for sorting
might not be requested by the user.

The PR also adds minor tweaks, mainly related to tree output.

Close #35118

costin committed:
SQL: Allow sorting of groups by aggregates

Introduce client-side, custom sorting of groups based on aggregate
functions. To allow this, the Analyzer has been extended to push down
to underlying Aggregate, aggregate function and the Querier has been
extended to identify the case and consume the results in order and sort
them based on the given columns.
The underlying QueryContainer has been slightly modified to allow a view
of the underlying values being extracted as the columns used for sorting
might not be requested by the user.

The PR also adds minor tweaks, mainly related to tree output.

Close #35118

elasticmachine commented:
Pinging @elastic/es-search

costin commented:
There are still some things that need to be added in a future PR, mainly:

1. move the metrics from the transport layer to the planExecutor so in case of sorting, the stats are properly incremented (since multiple requests are done).
2. figure out the policy for establishing the maximum number of items to be sorted. Currently the PR uses an arbitrary number (100) however this requires investigation.
While initially I was favoring the idea of the user being required to specify a limit - `LIMIT 100` I think that's quite annoying and in most cases we'll have an upper limit anyway.
So another variant is to always allocate the higher limit in case no LIMIT is specified and throw an exception at runtime if the limit is reached to tell the user that they have to specify a manual LIMIT.

costin committed:
Pass column size explicitly since the schema is lost on paging calls

matriv commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java:
```diff
@@ -778,31 +790,23 @@ protected LogicalPlan rule(Project project) {
-            // if the sort points to an agg, consider it only if there's grouping
-            if (ob.child() instanceof Aggregate) {
-                Aggregate a = (Aggregate) ob.child();
+            AtomicBoolean foundAggregate = new AtomicBoolean(false);
```
here too? why AtomicBoolean?

matriv commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plan/physical/EsQueryExec.java:
```diff
@@ -56,8 +54,21 @@ public QueryContainer queryContainer() {
-        Querier scroller = new Querier(session.client(), session.configuration());
-        scroller.query(Rows.schema(output), queryContainer, index, listener);
+        Querier scroller = new Querier(session);
+        //        List<ExpressionId> ids = queryContainer.columns();
```
Leftover?

matriv commented on x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/analysis/analyzer/VerifierErrorMessagesTests.java:
```diff
@@ -312,14 +312,19 @@ public void testUnsupportedTypeInOrder() {
-    public void testGroupByOrderByNonKey() {
-        assertEquals("1:52: Cannot order by non-grouped column [a], expected [bool]",
-                error("SELECT AVG(int) a FROM test GROUP BY bool ORDER BY a"));
+    //    public void testGroupByOrderByNonKey() {
```
leftover, please remove.

matriv reviewed (commented):
Looks good overall, Really important feature!
Left some comments and additionally:
1. Please add tests with ordering on multiple columns (tie breaker)
2. Please add tests with ordering using number indexes (e.g.: ORDER BY 1, 2)
3. Please add tests with descending ordering
4. Please remove the paragraph from the limitations doc page.

astefan reviewed (commented):
Great stuff. Left few minor comments

astefan commented:
```
sql> select gender as g, min(salary) as min, max(salary) as max from test_emp group by 1 order by 2, 3, 2, 4 desc limit 3;
Bad request [line 1:103: Invalid %d specified in OrderBy (valid range is [1, %d])]
```

astefan commented on x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/analysis/analyzer/VerifierErrorMessagesTests.java:
```diff
@@ -312,14 +312,19 @@ public void testUnsupportedTypeInOrder() {
-        assertEquals("1:44: Cannot order by non-grouped column [MAX(int)], expected [int]",
-                error("SELECT int FROM test GROUP BY int ORDER BY MAX(int)"));
+    public void testGroupByOrderByAggAndGroupedColumn() {
+        assertEquals("1:49: Cannot order by aggregated [MAX(int)] and non-aggregated [int] columns"
```
Can you add a test where `ORDER BY` has more than two elements?

matriv commented:
I would also add the `100` implicit limit to the docs.

costin commented on x-pack/plugin/sql/qa/src/main/resources/agg-ordering.sql-spec:
```diff
@@ -0,0 +1,51 @@
+aggWithAlias
+SELECT MAX(salary) AS m FROM test_emp GROUP BY gender ORDER BY m;
+
+multipleAggsThatGetRewrittenWithoutAlias
```
max and avg and count get rewritten/replaced with a Stats agg.

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Verifier.java:
```diff
@@ -310,11 +310,31 @@ private static boolean checkGroupByOrder(LogicalPlan p, Set<Failure> localFailur
                 Map<Expression, Node<?>> missing = new LinkedHashMap<>();
+
+                // track aggs and non-aggs - to keep the final modifier, use an array
+                final Expression[] aggAndNonAgg = new Expression[2];
```
No, because if found, the expression is used in the validation message.

matriv replied:
sorry, missed that part

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Analyzer.java:
```diff
@@ -1001,6 +999,67 @@ protected LogicalPlan rule(LogicalPlan plan) {
+
+            // 2. find first Aggregate child and update it
+
+            final AtomicBoolean found = new AtomicBoolean(false);
```
Not it's an utility class for storing state outside a lamba expression. Since the class is already part of the JDK (so it's likely to be inlined) it's quite convenient. The alternative is to create our own Holder class.

matriv replied:
got it thx!

matriv replied:
I would opt for a `boolean[] found = new boolean[] {false}` instead, to avoid confusion and because AtomicBoolean has special semantics for the compiler to achieve the atomic operations in CPU level which is a bit too much, imho, just for convenience.

costin commented on x-pack/plugin/sql/qa/single-node/src/test/java/org/elasticsearch/xpack/sql/qa/single_node/CliExplainIT.java:
```diff
@@ -64,22 +64,22 @@ public void testExplainWithWhere() throws IOException {
         assertThat(readLine(), startsWith("\\_Project[[?*]]"));
-        assertThat(readLine(), startsWith("  \\_Filter[i = 2#"));
-        assertThat(readLine(), startsWith("    \\_UnresolvedRelation[[][index=test],null,Unknown index [test]]"));
+        assertThat(readLine(), startsWith("  \\_Filter[Equals[?i"));
```
We can for unresolved items but for resolved ones not really since the tests are ran in random order and the ids change between runs.

matriv replied:
Not important at all but can't we use `endsWith`?

costin replied:
No because 2 is a literal with its own id.

costin commented on x-pack/plugin/sql/qa/src/main/resources/agg-ordering.sql-spec:
```diff
@@ -0,0 +1,51 @@
+// Custom sorting/ordering on aggregates
+//
+
+countWithImplicitGroupBy
```
Indeed but there's a dedicated rule in the optimizer for this case that excludes the order (as there's no physical group for it).

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Verifier.java:
```diff
@@ -310,11 +310,31 @@ private static boolean checkGroupByOrder(LogicalPlan p, Set<Failure> localFailur
                 Map<Expression, Node<?>> missing = new LinkedHashMap<>();
+
+                // track aggs and non-aggs - to keep the final modifier, use an array
+                final Expression[] aggAndNonAgg = new Expression[2];
```
I don't think so because one would have two arrays - this is really a hack for the setting the state outside a lamba expression. I'll probably go for a `Holder` class which be more meaningful in intent.

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/search/Querier.java:
```diff
@@ -95,30 +107,147 @@ public void query(Schema schema, QueryContainer query, String index, ActionListe
+    /**
+     * Listener used for local sorting (typically due to aggregations used inside `ORDER BY`).
+     * 
+     * This listener consumes the whole result set, sorts it in memory then send the paginates
```
"sends the paginated" response. Fixed.

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plan/physical/EsQueryExec.java:
```diff
@@ -56,8 +54,21 @@ public QueryContainer queryContainer() {
-        Querier scroller = new Querier(session.client(), session.configuration());
-        scroller.query(Rows.schema(output), queryContainer, index, listener);
+        Querier scroller = new Querier(session);
+        //        List<ExpressionId> ids = queryContainer.columns();
```
No, I've removed them

costin commented on x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/analysis/analyzer/VerifierErrorMessagesTests.java:
```diff
@@ -312,14 +312,19 @@ public void testUnsupportedTypeInOrder() {
                 error("SELECT * FROM test ORDER BY unsupported"));
     }
 
-    public void testGroupByOrderByNonKey() {
```
This needs a bit of reworking.

costin commented:
@matriv see number 2) in my initial comment.

matriv commented:
@costin yep read the comment, I was just thinking to have like an autonomous complete PR. But I'm totally fine to add the proper docs once this bit is figured out and implemented correctly. Personally, if no LIMIT is provided by the user, I'd favour to have a documented implicit limit rather than exhausting available memory and throw an Exception. On the other hand if the latter can implemented easily with the use of a CircuitBreaker maybe it's a nice option too.

costin committed:
First round of feedback

costin committed:
Merge remote-tracking branch 'remotes/upstream/master' into fix-35118

costin committed:
Address feedback

costin committed:
wip

costin committed:
Merge branch 'master' into fix-35118

costin committed:
wip

costin committed:
Merge branch 'master' into fix-35118

costin committed:
Address feedback

costin commented:
> ```
> sql> select gender as g, min(salary) as min, max(salary) as max from test_emp group by 1 order by 2, 3, 2, 4 desc limit 3;
> Bad request [line 1:103: Invalid %d specified in OrderBy (valid range is [1, %d])]
> ```

I've created a separate issue for this since it needs backporting to 6.x. It's not just the message but also the fact that the Analyzer rule throws an exception (with an incorrect message).
See https://github.com/elastic/elasticsearch/issues/38208

costin commented:
> @costin yep read the comment, I was just thinking to have like an autonomous complete PR. But I'm totally fine to add the proper docs once this bit is figured out and implemented correctly. Personally, if no LIMIT is provided by the user, I'd favour to have a documented implicit limit rather than exhausting available memory and throw an Exception. On the other hand if the latter can implemented easily with the use of a CircuitBreaker maybe it's a nice option too.

In case the LIMIT is not specified, the hard limit (256) is enforced however the moment an entry is dropped, an exception is thrown.
The idea being that we can sort 256 entries however more than that and we lose some records - and appropriately tell the user instead of silently applying that.

costin commented:
@astefan @matriv I've addressed more comments, beefed up the test suite (with some tie breakers) and addressed point 2).
There will be a follow-up PR regarding the move of metrics and one for the docs to add some examples.

matriv reviewed (approved):
LGTM!!!

astefan reviewed (approved):
LGTM. Great feature.

costin committed:
Tie breaker tests

costin committed:
Merge remote-tracking branch 'remotes/upstream/master' into fix-35118

costin commented:
@elasticmachine run elasticsearch-ci/packaging-sample
